src/main/scala/dibl/proto/TilesConfig.scala

```diff
@@ -83,56 +90,91 @@ import scala.scalajs.js.annotation.{ JSExport, JSExportTopLevel }
+          vectorCode,
+          stitch,
+          row < inputMatrix.length,
+          relativeSources = Matrix.toRelativeSources(vectorCode)
+        )
     }
   }
 
-  // repeat tiles, see: https://github.com/d-bl/GroundForge/blob/2e96d8b5/docs/help/images/shift-directions.png
-
-  for { // TODO reduce ranges to avoid if
-    i <- itemMatrix.indices
-    j <- -centerCols until centerCols
-    translateRow = (i * shiftRowsSE) + (j * shiftRowsSW)
-    translateCol = (i * shiftColsSE) + (j * shiftColsSW)
-    r <- centerMatrix.indices
-    c <- centerMatrix(r).indices
-  } {
-    // t in rt/ct stands for target cell, r and c for row and col
-    val rt = r + translateRow
-    val ct = c + translateCol
-    if (rt >= 0 && ct >= 0 && rt < totalRows && ct < centerCols) {
-      val id = Stitches.toID(r, c + leftMarginWidth)
-      val vectorCode = centerMatrix(r)(c)
-      val stitch = if ("-VWXYZ".contains(vectorCode.toUpper)) "-"
-                   else fields.getOrElse(id, centerMatrixStitch)
-      itemMatrix(rt)(ct + leftMarginWidth) = Item(id, vectorCode, stitch, r == rt && c == ct,
-        relativeSources = Matrix.toRelativeSources(vectorCode))
+  /**
+    * @param offsetOfFirstTile horizontal position of the original tile within the targetMatrix
+    * @param rows              height of the tile
+    * @param cols              width of the tile
+    * @param startRow          top position for the new tile within the targetMatrix
+    * @param startCol          left position for the new tile within the targetMatrix
+    */
+  private def copyTile(offsetOfFirstTile: Int, rows: Int, cols: Int, startRow: Int, startCol: Int): Unit = {
+    for {
+      row <- 0 until rows // row within the tile
+      col <- 0 until cols // col withing the tile
+    } {
+      val targetRow = startRow + row
+      val targetCol = startCol + col + offsetOfFirstTile
+      val sourceCol =  col + offsetOfFirstTile
+      val sourceRow =  row
+      if (targetCol >= 0 && targetCol < totalCols &&
+        targetRow >= 0 && targetRow < totalRows) {
+        targetMatrix(targetRow)(targetCol) = targetMatrix(sourceRow)(sourceCol).copy(isOpaque = false)
+      }
     }
   }
 
+  private def repeatSide(offsetOfFirstTile: Int, rows: Int, cols: Int): Unit = {
+    if (rows > 0 && cols > 0)
+      for {row <- rows until patchHeight by rows} {
+        copyTile(offsetOfFirstTile, rows, cols, startRow = row, startCol = 0)
+      }
+  }
+
+  private def copyCenterTile(startRow: Int, startCol: Int): Unit =
+    copyTile(leftMatrixCols, centerMatrixRows, centerMatrixCols, startRow, startCol)
+
+  setFirstTile(centerMatrix, leftMarginWidth, centerMatrixStitch)
+  if (centerMatrixRows > 0 && centerMatrixCols > 0 && patchWidth > 0 && patchHeight > 0) {
+    // TODO concise loop boundaries
+    val squaredPatchSize = Math.max(patchWidth, patchHeight)
+    for {i <- 0 until squaredPatchSize} {
+      for {j <- squaredPatchSize until -squaredPatchSize by -1} if (!(i==0 & j==0)) {
```

Comment:
I suppose we can reduce the range for `j` when the the difference between `patchWidth` and `patchHeight` is larger than the tile and/or shift values.